src/peft/tuners/bone/layer.py

```diff
@@ -0,0 +1,260 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import math
+import warnings
+from typing import Any, List, Optional, Union
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from peft.tuners.tuners_utils import BaseTunerLayer, check_adapters_to_merge
+
+from einops import rearrange
+
+
+class BoneLayer(BaseTunerLayer):
+    # All names of layers that may contain (trainable) adapter weights
+    adapter_layer_names = ("bone_block",)
+    # All names of other parameters that may contain adapter-related parameters
+    other_param_names = ("bone_r")
+
+    def __init__(self, base_layer: nn.Module, **kwargs) -> None:
+        self.base_layer = base_layer
+        self.bone_r = {}
+        self.bone_block = nn.ParameterDict({})
+        # Mark the weight as unmerged
+        self._disable_adapters = False
+        self.merged_adapters = []
+        self.kwargs = kwargs
+
+        base_layer = self.get_base_layer()
+        if isinstance(base_layer, nn.Linear):
+            self.in_features, self.out_features = base_layer.in_features, base_layer.out_features
+        elif isinstance(base_layer, nn.Conv2d):
+            self.in_features, self.out_features = base_layer.in_channels, base_layer.out_channels
+        else:
+            raise ValueError(f"Unsupported layer type {type(base_layer)}")
+
+    def update_layer(
+        self,
+        adapter_name: str,
+        r: int,
+        init_weights: bool,
+        **kwargs,
+    ) -> None:
+        """Internal function to create bone adapter
+
+        Args:
+            adapter_name (`str`): Name for the adapter to add.
+            r (`int`): Rank for the added adapter.
+            init_weights (`bool`): Whether to initialize weights.
+            apply_GS (`bool`): Whether to apply Gram-Schmidt orthogonalization or not.
+        """
+        if r <= 0:
+            raise ValueError(f"`r` should be a positive integer value but the value passed is {r}")
+
+        self.bone_r[adapter_name] = r
+
+        # Determine shape of Bone weights
+        base_layer = self.get_base_layer()
+        if isinstance(base_layer, nn.Linear):
+            self.bone_block[adapter_name] = nn.Parameter(torch.zeros(self.out_features//r, r, r), requires_grad=True)
+        # elif isinstance(base_layer, nn.Conv2d):
+        #     self.bone_block[adapter_name] = nn.Parameter(
+        #         torch.empty(self.in_features * base_layer.kernel_size[0] * base_layer.kernel_size[0], r),
+        #         requires_grad=True,
+        #     )
+        else:
+            raise TypeError(f"Bone is not implemented for base layers of type {type(base_layer).__name__}")
+
+        # Initialize weights
+    #     if init_weights:
```

Comment:
Should this code be uncommented? As is, `init_weights` has no effect.

Reply:
I think Bone does not need to init_weight

Reply:
Just to explain, we usually have the `init_weight` parameter so that we can switch between initializing the weights as identity transform (default, as it normally helps training) and non-identity transform (non-default). Having a non-identity initialization is nice to have for unit testing, where we want to measure the effect of the adapter.

Reply:
Did you consider this?

Reply:
The init_weight is now provided, but it is still initialized from 0.